PicPi/FrameAPI.swift

```diff
@@ -0,0 +1,47 @@
+//
+//  FrameAPI.swift
+//  PicPi
+//
+//  Created by Mustafa T. Mohammed on 12/21/20.
+//
+
+import Foundation
+import Promises
+
+func injectFrameAPI() -> FrameAPI {
+  return FrameAPIImpl()
+}
+/**
+ # Example usage
+ 
+ ## Initialization
+   ```
+  let frameAPI: FrameAPI
+  init(frameAPI: FrameAPI = injectFrameAPI()) {
+   self.frameAPI = frameAPI
+  }
+ ```
+## Call function that return for example [URL]
+ - Notice that we use (.then) that will return the list of URLs and (.catch) that will return the rejection reason (error)
+ ```
+  func getThumbnails() {
+   frameAPI.retrieveThumbnails().then {  Urls    in
+     print(Urls)`
+   }.catch {reject in
+     print(reject)
+    }
+  }
+ ```
+ */
+protocol FrameAPI {
+  /// Upload the selected photo with filename
+  func upload(photoData: Data, filename: String) -> Promise<Void>
+  /// Get the list of URLs of the thumbnails stored on the Pi
+  func retrieveThumbnails() -> Promise<[URL]>
+  /// Remove the given photo
+  func removePhoto(filename: String) -> Promise<Void>
+  /// Get Pi Preferences
+  func retrievePIPrefrences() -> Promise<[String : Any]>
+  /// Update Pi Preferences 
+  func updatePI(preferences: [String : Any]) -> Promise<Void>
```

Comment:
Wdyt about using a codable struct for the preferences instead of a dictionary? That way when I update the preferences I don't send garbage keys nor have to type check the values

Reply:
Great idea,
I see that you have added this in PictureFrame.swift:
struct PictureFramePreferences {
  let slideshowDuration: TimeInterval
  let connectionPasscode: String
}
do you want me to use it and make it Codable?

Reply:
Not sure what's best since I haven't worked with mutable objects in the UI yet but wdyt of this

Since the http request is patch (meaning it'll only store whatever is present in the json body) maybe we can create another struct with optional vars for each preference like so:

Reply:
Spoke offline, just a discussion to not use Codable in user object. Codable is good for now